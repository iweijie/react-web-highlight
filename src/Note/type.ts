import { INoteTextHighlightInfoItem } from '../getSelectedInfo';

export interface IModeProps {
  mode: string;
  className?: string;
}

export interface INoteTextHighlightInfo {
  list: INoteTextHighlightInfoItem[];
  text: string;
  mode?: string;
  [x: string]: any;
}

export interface INote {
  template: string;
  value?: INoteTextHighlightInfo[];
  tagName?: string;
  // attrName?: string;
  // splitAttrName?: string;
  onAdd?: (props: INoteTextHighlightInfo) => void;
  onUpdate?: (props: INoteTextHighlightInfo[]) => void;
  rowKey?: string;
  modes?: IModeProps[];
}
