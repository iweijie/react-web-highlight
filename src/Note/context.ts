import { createContext } from 'react';
import { INoteTextHighlightInfo } from './type';
export interface INoteContextProps {
  wrapContainer: React.RefObject<HTMLDivElement>;
  selectedValue: React.RefObject<INoteTextHighlightInfo | null>;
  action: React.RefObject<string>;
}

export default createContext<INoteContextProps | null>(null);
