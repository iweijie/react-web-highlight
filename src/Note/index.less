[data-wj-custom-split] {
  cursor: pointer;
}

.text-highlight-wrap {
  position: relative;
}
