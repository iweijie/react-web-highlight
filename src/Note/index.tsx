import React, {
  useState,
  useCallback,
  useMemo,
  useRef,
  FC,
  useEffect,
} from 'react';
import useUpdateEffect from '../hooks/useUpdateEffect';
import getSelectedInfo from '../getSelectedInfo';
import {
  customTag as cTag,
  customAttr as cAttr,
  customRowKey,
  customSplitAttr,
  customSelectedAttr,
} from '../constants';
import { setCustomValue } from '../customAttrValue';
import Context from './context';
import { INote, INoteTextHighlightInfo } from './type';
import Parse from '../Parse/index';
import type {IAstItem} from '../Parse/index';
import './index.less';
import { isEmpty } from 'lodash';


const getContext = (snapShoot:IAstItem[])=>{

  const nodes: any = []

  for(let i = 0; i <snapShoot.length; i++) {
    const node = snapShoot[i]
    if(node.type === 'text') {
      nodes.push(node.content)
    }else if(node.type === 'element') {
      let children = []
      if(!isEmpty(node.children)) {
         children = getContext(node.children)
      }
      const props = node.attributes.reduce((attrs: any, attribute)=>{

        if(attribute.name === 'class') {
          attrs.className = attribute.value
        }else {
          
          attrs[attribute.name] = attribute.value
        }


        return attrs
      }, {})

      props.key = `${node.tagName}-${i}`

      nodes.push(React.createElement(node.tagName, props, ...children ))
    }else {
      throw new Error("类型错误")
    }

  }

  return nodes
}

const Note: FC<INote> = ({
  template,
  value,
  tagName = cTag,
  rowKey = customRowKey,
  onAdd,
  onUpdate,
  modes,
  children,
}) => {
  const parse = useMemo(() => {
    // 用于格式化html文本
    if (template) {
      const div = document.createElement('div');
      div.innerHTML = template;
      template = div.innerHTML;
    }
    return new Parse({ template });
  }, []);

  const [snapShoot, setSnapShoot] = useState<IAstItem[]>(() => {
    return parse.getJSON(value);
  });
  const selectedValue = useRef<INoteTextHighlightInfo | null>(null);
  const action = useRef<string>('');
  const noteContainer = useRef<HTMLDivElement>(null);
  const wrapContainer = useRef<HTMLDivElement>(null);

  const [contextValue, setContextValue] = useState({
    selectedValue,
    wrapContainer,
    action,
  })
  const handleSelectedText = useCallback(() => {
    const range: Range | undefined = window?.getSelection()?.getRangeAt(0);
    if (!range) return;

    const text = range.toString();

    const { collapsed = true, endContainer, startContainer } = range;

    // 返回条件 1. 光标起始点相同（即没有选中文本），2. 起点或者终点不在当前容器内
    if (
      collapsed ||
      !noteContainer.current ||
      !noteContainer.current.contains(startContainer as Node) ||
      !noteContainer.current.contains(endContainer as Node)
    )
      return;

    const list = getSelectedInfo({
      range: range as Range,
      noteContainer: noteContainer.current,
    });

    const data = {
      list,
      text: text,
    };

    action.current = 'add';
    selectedValue.current = data;

    onAdd && onAdd(data);
  }, [noteContainer, modes, parse]);

  const handleClick = useCallback(
    e => {
      if (!noteContainer.current || !e.target) return;
      let node = e.target;
      const uuids = [];
      while (node) {
        const uuid = node.getAttribute(customSelectedAttr);
        if (uuid) {
          uuids.push(uuid);
          node = node.parentNode;
        } else {
          break;
        }
      }

      if (!uuids.length) return;

      const findData: INoteTextHighlightInfo[] = [];

      uuids.forEach(uuid => {
        const data = value?.find(item => item[rowKey] === uuid);
        if (data) {
          findData.push(data);
        }
      });

      action.current = 'update';
      onUpdate && onUpdate(findData);
    },
    [noteContainer, value]
  );

  const handleMouseDown = useCallback(
    eventDown => {
      if (!noteContainer.current) return;
      document.addEventListener(
        'mouseup',
        eventUp => {
          handleSelectedText();
        },
        {
          once: true,
        }
      );
    },
    [noteContainer]
  );

  useEffect(()=>{
    setContextValue({
      selectedValue, action, wrapContainer
    })
  }, [selectedValue, action, wrapContainer]);

  useEffect(() => {
    // TODO 不造是否有更好的设置值的方式，后续如果有再修改
    const modeClassNames: any = {};
    if (modes && modes.length) {
      modes.forEach(item => {
        if (item.mode) {
          modeClassNames[item.mode] = item.className || '';
        }
      });
    }
    setCustomValue({
      tagName,
      attrName: cAttr,
      rowKey,
      splitAttrName: customSplitAttr,
      selectedAttr: customSelectedAttr,
      modeClassNames,
    });
    setSnapShoot( parse.getJSON(value) );
  }, [setSnapShoot, parse, value, modes]);

  return (
    <Context.Provider value={contextValue}>
      <div className="text-highlight-wrap" ref={wrapContainer}>
        <div
          className="text-highlight"
          id="text-highlight"
          ref={noteContainer}
          onMouseDown={handleMouseDown}
          onClick={handleClick}
        >
          {getContext(snapShoot)}
        </div>
        {children}
      </div>
    </Context.Provider>
  );
};

export default Note;
