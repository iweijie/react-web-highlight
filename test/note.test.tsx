import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Note from '../src/Note';

const getHtml = (html: string): string => {
  return `<div class="text-highlight-wrap"><div class="text-highlight" id="text-highlight">${html}</div></div>`;
};

describe('测试划线功能', () => {
  const template = '<p>我是很长很长很长的文本</p>';

  const modes = [
    {
      className: 'huaxian',
      mode: 'huaxian',
      name: '划线',
    },
    {
      className: 'edit',
      mode: 'edit',
      name: '笔记',
    },
  ];

  it('basic', () => {
    const wrapper = render(<Note modes={modes} template={template} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('text', () => {
    const app = render(<Note modes={modes} template={template} />);
    expect(app.text()).toEqual('我是很长很长很长的文本');
  });

  it('单一下划线', () => {
    const value = [
      {
        list: [{ level: [0, 0], start: 2, end: 8, text: '很长很长很长' }],
        text: '很长很长很长',
        mode: 'huaxian',
        id: '8f5e0551a588f1',
      },
    ];

    const app = mount(<Note modes={modes} template={template} value={value} />);

    expect(app.find('.huaxian').text()).toEqual('很长很长很长');

    expect(app.html()).toEqual(
      getHtml(
        '<p>' +
          '<span data-wj-custom-text="true">' +
          '我是' +
          '<span data-wj-custom-split="true" data-wj-custom-id="8f5e0551a588f1" class="huaxian">很长很长很长</span>' +
          '的文本' +
          '</span>' +
          '</p>'
      )
    );

    expect(toJson(app)).toMatchSnapshot();
  });

  it('多下划线', () => {
    const value = [
      {
        list: [{ level: [0, 0], start: 2, end: 8, text: '很长很长很长' }],
        text: '很长很长很长',
        mode: 'huaxian',
        id: '8f5e0551a588f1',
      },
      {
        list: [{ level: [0, 0], start: 3, end: 5, text: '长很' }],
        text: '长很',
        mode: 'huaxian',
        id: '8f5e0551a588f2',
      },
    ];

    const app = mount(<Note modes={modes} template={template} value={value} />);
    app.update();

    expect(
      app
        .find('[data-wj-custom-id="8f5e0551a588f1"]')
        .find('[data-wj-custom-id="8f5e0551a588f2"]')
        .text()
    ).toEqual('长很');

    expect(app.find('[data-wj-custom-id="8f5e0551a588f1"]').length).toBe(3);
    
    const testText = ['很', '长很', '长很长']
    app.find('[data-wj-custom-id="8f5e0551a588f1"]').forEach((node, index) => {
      expect(node.text()).toEqual(testText[index]);
    });

    expect(app.html()).toEqual(
      getHtml(
        '<p>' +
          '<span data-wj-custom-text="true">' +
          '我是' +
          '<span class="huaxian" data-wj-custom-split="true" data-wj-custom-id="8f5e0551a588f1">很</span>' +
          '<span class="huaxian" data-wj-custom-split="true" data-wj-custom-id="8f5e0551a588f1">' +
          '<span class="huaxian" data-wj-custom-split="true" data-wj-custom-id="8f5e0551a588f2">' +
          '长很' +
          '</span>' +
          '</span>' +
          '<span class="huaxian" data-wj-custom-split="true" data-wj-custom-id="8f5e0551a588f1">长很长</span>' +
          '的文本' +
          '</span>' +
          '</p>'
      )
    );

    expect(toJson(app)).toMatchSnapshot();
  });

  it('测试修改props，dom结构更新', () => {
    const value = [
      {
        list: [{ level: [0, 0], start: 2, end: 8, text: '很长很长很长' }],
        text: '很长很长很长',
        mode: 'huaxian',
        id: '8f5e0551a588f1',
      },
    ];

    const app = mount(<Note modes={modes} template={template} />);
    app.update();

    expect(app.find('.huaxian').length).toBe(0);

    app.setProps({
      value,
    });
    app.update();

    expect(app.find('.huaxian').length).toBe(1);

    expect(app.find('.huaxian').text()).toEqual('很长很长很长');
  });
});
