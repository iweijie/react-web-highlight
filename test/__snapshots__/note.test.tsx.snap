// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试划线功能 basic use 1`] = `
<div
  class="text-highlight-wrap"
>
  <div
    class="text-highlight"
    id="text-highlight"
  />
</div>
`;
